File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        res = 0
        
        def dfs(root: Optional[TreeNode], dir: int, cur: int) -> None:
            nonlocal res
            if not root:
                return
            res = max(res, cur)
            dfs(root.left, -1, 1 + (cur if dir == 1 else 0))
            dfs(root.right, 1, 1 + (cur if dir == -1 else 0))
        
        dfs(root, 0, 0)
        return res

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
    let res = 0
    dfs(root, 0, 0)
    return res
    
    function dfs(root, dir, cur) {
        if (!root) return
        res = Math.max(res, cur)
        dfs(root.left, -1, 1 + (dir === 1 ? cur : 0))
        dfs(root.right, 1, 1 + (dir === -1 ? cur : 0))
    }
};